qcodes/instrument/base.py

```diff
@@ -17,6 +18,22 @@
 
 log = logging.getLogger(__name__)
 
+class InstrumentLoggerAdapter(logging.LoggerAdapter):
+    def process(self, msg, kwargs):
+        kwargs['extra'] = self.extra
+        return f"[{self.extra['instrument']}] {msg}", kwargs
+
+class InstrumentFilter(logging.Filter):
+    def __init__(self, instruments):
+        if isinstance(instruments, InstrumentBase):
+            instruments = (instruments,)
+        self.names = [instrument.full_name for instrument in instruments]
+
+    def filter(self, record):
+        try:
+            return record.instrument in self.names
```

Comment:
The extra keyword is something that comes from the python logging module. But I can add the comment here as well if you want.

Reply:
no no, the `extra` keyword itself is ok and is documented in python, sure. The important part is that if one wants to use an instrument adapter/filter, then one needs to pass this dict as "extra". otherwise the filter will not work, right? so, this info is missing in the docstring somewhere.

Reply:
You need to pass it to `LoggerAdapter.__init__` documented here:
https://docs.python.org/3.6/library/logging.html?highlight=loggeradapter#logging.LoggerAdapter
I added a docstring for clarity.

Reply:
great! now it is clear how to create the InstrumentLoggerAdapter so that it works with the InstrumentFilter.

btw, is it a good idea to hide this detail into the InstrumentLoggerAdapter constructor, for example, like this:
```
class InstrumentLoggerAdapter(logging.LoggerAdapter):
    def __init__(self, instrument_instance, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.extra.update({'instrument': self.full_name})
```
(perhaps a better way is to modify kwargs before calling super.__init__)

Reply:
Your example is missing the required `extra` argument to the `LoggerAdapter.__init__`-call.

The `InstrumentLoggerAdapter` is supposed to  be used in a single place, which is the `BaseInstrument.__init__` method. I will therefore not deviate from the default way `LoggerAdapter`s are initialized.

Reply:
oh, ok, understand. i missed the part that `extra` is a required kwarg.

Reply:
Sorry for possibly misunderstanding your point:
When logging from an instrument as it is of now, using `self.log`, the records will be logged to the `InstrumentBase` logger. 
I changed the code so that every instrument now has: `self.log_base`, `self.log_visa` and `self.log`. The reason for having the extra logs around is that I discovered, that in some drivers the write_raw function gets overwritten and the visa logger is used.
I am not sure if this is the most elegant mechanism, but it mimicks what is present already.

Reply:
my point was about smth else, but let's skip it, it's ok.

your new point about "in some drivers the write_raw function gets overwritten and the visa logger is used" - very valid point. it's tricky here because one can log from the perspective of a module or from an object instance. for the case of instruments, we want to log from the perspective of the instrument instance in order to be able to filter by that instance (and add `[<instrument name>]` prefix to log messages). However, when one implements a driver and wants to use logging, the typical pattern is to declare `log = logging.getLogger(__name__)` at the top of the module. 

I don't have any suggestions at the moment... but i think that we can define whatever works as long as we stick to it and we document it in the "how to write a driver" notebook.